lib/schema/src/validate.rs

```diff
@@ -751,3 +904,304 @@ impl RegisterBlock {
+                }
+                .into(),
+                Register {
+                    name: "data2".into(),
+                    offset: 24,
+                    array_dimensions: vec![16],
+                    ..Default::default()
+                }
+                .into(),
+            ],
+            ..Default::default()
+        };
+        let mut block = block.validate_and_dedup().unwrap();
+        block.extract_array("data").unwrap();
+        assert_eq!(
+            block.block,
+            RegisterBlock {
+                name: "block".into(),
+                registers: vec![
+                    Register {
+                        name: "unrelated0".into(),
+                        offset: 0,
+                        ..Default::default()
+                    }
+                    .into(),
+                    Register {
+                        name: "data".into(),
+                        offset: 16,
+                        array_dimensions: vec![3, 16,],
+                        ..Default::default()
+                    }
+                    .into(),
+                ],
+                ..Default::default()
+            }
+        );
+    }
+
+    #[test]
+    fn test_extract_array_dimension_mismatch() {
+        let block = RegisterBlock {
+            name: "block".into(),
+            registers: vec![
+                Register {
+                    name: "data0".into(),
+                    offset: 16,
+                    ..Default::default()
+                }
+                .into(),
+                Register {
+                    name: "data1".into(),
+                    offset: 20,
+                    array_dimensions: vec![2],
+                    ..Default::default()
+                }
+                .into(),
+            ],
+            ..Default::default()
+        };
+        let mut block = block.validate_and_dedup().unwrap();
+        assert_eq!(
+            block.extract_array("data").unwrap_err().to_string(),
+            "Inconsistent register array dimensions in data1"
+        );
+    }
+
+    #[test]
+    fn test_extract_array_type_mismatch() {
+        let block = RegisterBlock {
+            name: "block".into(),
+            registers: vec![
+                Register {
+                    name: "data0".into(),
+                    offset: 16,
+                    ty: RegisterType {
+                        fields: vec![RegisterField {
+                            name: "foo".into(),
+                            width: 5,
+                            ..Default::default()
+                        }
+                        .into()],
+                        ..Default::default()
+                    }
+                    .into(),
+                    ..Default::default()
+                }
+                .into(),
+                Register {
+                    name: "data1".into(),
+                    offset: 20,
+                    array_dimensions: vec![2],
```

Comment:
nit: this test would also fail because of `ValidationError::RegisterArrayInconsistentDimensions`. If we just wanted to test the type mismatch, I'd remove this mismatched dimension.

Reply:
Good idea. Done.